xoto3/dynamodb/write_versioned/README.md

```diff
@@ -0,0 +1,118 @@
+# versioned_transact_write_items
+
+This is a general-purpose system for writing retrying DynamoDB
+transactional operations for 1 to
+[25](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-dynamodb-transactions)
+items, using pure functions.
+
+The magic sauce here is the versioning - we use ConditionExpressions
+based on a specified attribute, by default `item_version`, to detect
+whether any intervening writes have been made, such that an `update`
+can safely be expressed as a simple `put`. This makes the
+[ABA problem](https://en.wikipedia.org/wiki/ABA_problem) impossible,
+so you can write your business logic free of worry about the
+complexity of transactions in the real world.
+
+The API follows a consistent pattern of using provided functions to
+modify an opaque, immutable VersionedTransaction object. All writes to
+the database are performed by calling a function with the current
+state of the versioned transaction, resulting in a new
+VersionedTransaction object that may be further evolved using other
+provided functions. Crucially, the API exposes no 'state' about the
+transaction except for via simple read and modify operations,
+resulting in a fully composable transaction system - any transaction
+builder can be composed with any other in whatever logical order is
+necessary.
+
+This API provides partial-application-by-default, because that makes
+certain use cases much cleaner to express. This means that you'll be
+writing things like `table.get(key)(transaction)` instead of
+`table.get(key, transaction)`, i.e. using more parentheses rather than
+commas.
+
+### put unless exists
+
+As a simple but non-trivial example, you can perform a put-unless-exists like so:
+
+```python
+import xoto3.dynamodb.write_versioned as wv
+
+# construct these at a module level and share them - they are
+# threadsafe because they are pure values.
+task_table = wv.ItemTable('TaskTable', item_name='Task')
```

Comment:
yeah, i think there's a place for that, but it's not here. This is an open source library, and we need to demonstrate the API, not advertise our other libraries.

incidentally, it would be preferable to use unstruc_strip_defaults:

`(lambda: TaskTable().name, Task.struc, unstruc_strip_defaults, "Task")`  - but this will be obvious soon enough from the examples that will be littered around the Stratus codebase.

I am not opposed to including something demonstrating how you'd use TypedTable that isn't coupled to `typecats`, but I couldn't come up with an easy-to-understand concept that didn't clutter the example with things that would prevent a new reader from focusing on the transaction parts.

Reply:
That makes sense. I almost didn't say anything, because I know this is pretty specific to how we'll be using this. I suppose we could put an example like this somewhere in stratus.